Add extractor for VrtNU

mrBliss opened the pull request:
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This PR address #11829 by adding a new extractor for VrtNU to `vrt.py`. The extractor is based on the existing extractor for canvas. As all videos require authentication the sole test is skipped. I have successfully used this extractor for a number of different videos (using `--cookies` for authentication).

I welcome all feedback.

mrBliss commented:
I don't mind squashing the commits together if that is preferred.

dstftw commented:
This is the copy paste of `CanvasIE`. You should generalize common code instead of duplicating it.

mrBliss commented:
True, all the code about `formats` and `subtitles` is copied from `CanvasIE`. However, the extraction of the info from the webpage is new. I thought at first about extending `CanvasIE`, but that would lead to many special cases for VrtNU, which does not help the readability. 

So how should the common code be generalised? By subclassing or by extracting it into methods that are put in `common.py` or in `canvas.py` as static methods or functions?

To me the two candidates for extraction seem to be:

```python
formats = []
for target in data['targetUrls']:
    format_url, format_type = target.get('url'), target.get('type')
    if not format_url or not format_type:
        continue
    if format_type == 'HLS':
        formats.extend(self._extract_m3u8_formats(
            format_url, display_id, entry_protocol='m3u8_native',
            ext='mp4', preference=0, fatal=False, m3u8_id=format_type))
    elif format_type == 'HDS':
        formats.extend(self._extract_f4m_formats(
            format_url, display_id, f4m_id=format_type, fatal=False))
    elif format_type == 'MPEG_DASH':
        formats.extend(self._extract_mpd_formats(
            format_url, display_id, mpd_id=format_type, fatal=False))
    else:
        formats.append({
            'format_id': format_type,
            'url': format_url,
        })
self._sort_formats(formats)
```
and:
```python
subtitles = {}
subtitle_urls = data.get('subtitleUrls')
if isinstance(subtitle_urls, list):
    for subtitle in subtitle_urls:
        subtitle_url = subtitle.get('url')
        if subtitle_url and subtitle.get('type') == 'CLOSED':
            subtitles.setdefault('nl', []).append({'url': subtitle_url})
```

Another hard problem to solve is to come up with sensible names for these methods :smiley:

dstftw commented:
You should not pollute `common.py` with particular extractor specific code. Easiest way is to extract common code in a base class.

mrBliss commented:
Another try, but now as a subclass of `CanvasIE`.

dstftw commented on youtube_dl/extractor/canvas.py:
```diff
@@ -117,3 +125,63 @@ def _real_extract(self, url):
+
+        data = self._download_json(
+            'https://mediazone.vrt.be/api/v1/vrtvideo/assets/%s'
+            % video_id, display_id)
```
170-172 - code duplication.

mrBliss replied:
Done

dstftw commented on youtube_dl/extractor/canvas.py:
```diff
@@ -117,3 +125,63 @@ def _real_extract(self, url):
+            'duration': float_or_none(data.get('duration'), 1000),
+            'thumbnail': data.get('posterImageUrl'),
+            'subtitles': subtitles,
+        }
```
178-187 - code duplication.

mrBliss replied:
Should I go for a method returning a tuple of `duration` and `thumbnail`? The SLOC will only increase though.

dstftw replied:
You should have only one single method for extracting info.

mrBliss replied:
Okay, I have now taken this approach and rebased again.

mrBliss replied:
@dstftw Have you had time to look at my latest changes yet? Thanks

rubenv commented:
I've been testing this PR and it works great. The Travis failure seems to be related to a timeout on the Travis side of things and should probably be ignored (ore restarted).

Nice work @mrBliss!

MichaelDeBoey commented:
Don't know if this can be merged @dstftw, but for what I'm seeing in this repo the Travis CI error(s) aren't related to @mrBliss' code

MichaelDeBoey commented:
I don't know how you implemented this @mrBliss, but I've found a really easy way of getting the correct download-link for VrtNU :-)

1. You just have to add `.securevideo.json` to the link (`https://www.vrt.be/vrtnu/a-z/thuis/22/thuis-s22a4160` will be `https://www.vrt.be/vrtnu/a-z/thuis/22/thuis-s22a4160.securevideo.json`)
2. You'll get an `mzid` from that `.json`-file.
3. Use that `mzid` at the end of `https://mediazone.vrt.be/api/v1/vrtvideo/assets/` so you get: `https://mediazone.vrt.be/api/v1/vrtvideo/assets/md-ast-0fa95ed5-4541-4e3f-b436-ee552bd613fb_1488278903849`
4. Now you see all sorts (`hls`, `.mp4`, ...) of download possibilities

mrBliss commented:
@MichaelDeBoey that's exactly how I've implemented it :smiley:

MichaelDeBoey commented:
> @MichaelDeBoey that's exactly how I've implemented it 😃
OK nice :-)
I just found out this method on userbase (the forum where I also found the method for #11912), so wanted to let you know :-)

MichaelDeBoey commented:
Any progress on this @mrBliss @dstftw?

Invictaz commented:
@mrBliss I'm running your version and it does not work. See my screenshot

-Running Python 3.6.0

`C:\Users\Jurgen\Downloads\youtube-dl-vrtnu\youtube-dl-vrtnu\youtube_dl>python __
main__.py https://www.vrt.be/vrtnu/a-z/postbus-x/3/postbus-x-s3a8
[VrtNU] postbus-x-s3a8: Downloading webpage
[VrtNU] postbus-x-s3a8: Downloading JSON metadata
ERROR: Unable to download JSON metadata: HTTP Error 401: Unauthorized (caused by
 <HTTPError 401: 'Unauthorized'>); please report this issue on https://yt-dl.org
/bug . Make sure you are using the latest version; see  https://yt-dl.org/update
  on how to update. Be sure to call youtube-dl with the --verbose flag and inclu
de its complete output.`

mrBliss commented:
@Invictaz the extractor doesn't yet support authentication, e.g., via `--username` and `--password`, so you are supposed to get that error.

My current workaround is to log in using my browser, export my cookies using the [Export Cookies add-on](https://addons.mozilla.org/en-gb/firefox/addon/export-cookies/) (a similar extension for Chrome undoubtedly exists), and pass them to youtube-dl via the `--cookies` flag. This has worked flawlessly for me for the past couple of months.

When I have some more time, I will try to add authentication.

Invictaz commented:
@mrBliss it's working now. Not the easy method I hoped for but undoubtly that will come. Please include in your script a method to extract the .vtt subtitles and convert them to .srt. This is in for example https://www.vrt.be/vrtnu/a-z/bobbejaan/2015/bobbejaan-s2015/

They are in .vtt format, very similar to SRT. In the .securevideo.json is also the link to the subtitles.

Do not make them hardcoded but as a separate file please.

Example of your script working:

C:\Users\Downloads\youtube-dl-vrtnu\youtube-dl-vrtnu\youtube_dl>python __
main__.py --cookies cookies.txt https://www.vrt.be/vrtnu/a-z/postbus-x/3/postbus
-x-s3a8
[VrtNU] postbus-x-s3a8: Downloading webpage
[VrtNU] postbus-x-s3a8: Downloading JSON metadata
[VrtNU] postbus-x-s3a8: Downloading JSON metadata
[VrtNU] postbus-x-s3a8: Downloading m3u8 information
[VrtNU] postbus-x-s3a8: Downloading f4m manifest
[VrtNU] postbus-x-s3a8: Downloading MPD manifest
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 124
[download] Destination: De pyromaan-mz-ast-6a65c311-e8a0-4be8-a4ad-7ff0ee9a25b6_
1467042023805.mp4
[download]  84.7% of ~240.25MiB at  1.02MiB/s ETA 00:51

Invictaz commented:
@mrBliss Oh and there is this

WARNING: mz-ast-6a65c311-e8a0-4be8-a4ad-7ff0ee9a25b6_1467042023805: malformated
aac bitstream. Install ffmpeg or avconv to fix this automatically.

mrBliss commented:
@Invictaz 
My extractor already downloads the .vtt files if you pass the right flag, but I don't think it's the responsibility of the extractor to convert them to .srt. You can just use `ffmpeg` for this: `ffmpeg -i foo.vtt foo.srt`.

That warning you get is normal and not specific to my extractor. Just read the full warning, it literally says how you can fix it.

mrBliss commented:
I added support for authentication via the `-u` and `-p` flags. Only e-mail accounts are supported though.

@MichaelDeBoey @Invictaz can you guys try this out with your accounts? Thanks

rubenv commented:
@mrBliss Works like a charm, good stuff! 👍

MichaelDeBoey commented:
@dstftw I think this one can be merged?

midas02 commented:
@mrBliss Can you check the following? I tried https://github.com/rg3/youtube-dl/pull/11873/commits/57fc66764242654892928a69d444467c7586fc95 (simplify login logic), and couldn't download http://www.vrt.be/vrtnu/a-z/reizen-waes/3/reizen-waes-s3a8-venezuela/. When accessing the link over httpS://, it did work! Tried release https://github.com/rg3/youtube-dl/pull/11873/commits/8317eeb261d1a69bc3db2493b1ec620191699c92, and that one didn't have the problem. But maybe I just made a mistake somewhere, please check.

mrBliss commented:
@midas02 The S in httpS is not the problem. The authentication seems to be a bit flaky, sometimes it just works when you try again. I don't think my last commit is the cause because I have encountered the issue before. For me it works 8 out of 10 times.

I'm not sure how to fix this, because I can't reliably reproduce it. Maybe it's some kind of a timing issue? Maybe I need to add a retry somewhere in case of a failed authentication.

midas02 commented:
@mrBliss Ok. I checked again, and it did work. False alarm.

MichaelDeBoey commented:
@dstftw I think this one can be merged?

MichaelDeBoey commented:
Any progress on this one?

CC: @mrBliss @dstftw

dstftw commented on youtube_dl/extractor/canvas.py:
```diff
@@ -110,10 +104,153 @@ def _real_extract(self, url):
+
+        title = self._html_search_regex(
+            r'<h1 class="content__heading">(.+?)</h1>',
+            webpage, 'title', default=None, flags=(re.M | re.S)).strip()
```
Consistently use flags inside regex.

dstftw commented on youtube_dl/extractor/canvas.py:
```diff
@@ -110,10 +104,153 @@ def _real_extract(self, url):
+        season_number = int_or_none(season)
+
+        # Sometimes the season is the year, so don't use it as the season number
+        if season_number and season_number > 1000:
```
This can be expressed in regex.

mrBliss replied:
If I do this with a regex and the season is something like for example "2017", both `season` and `season_number` will be `None`. Whereas previously, `season` would still be `"2017"`. For me the new behaviour is fine, as I don't care about non-numbered seasons.

dstftw commented on youtube_dl/extractor/canvas.py:
```diff
@@ -110,10 +104,153 @@ def _real_extract(self, url):
+                'duration': 3011.17,
+                'thumbnail': r're:^https?://.*\.jpg$'
+            },
+            'skip': 'This video is only available for registered users'
```
Plays fine without any authentication in browser.

mrBliss replied:
I cannot reproduce this, I have to authenticate in order to view [the video](https://www.vrt.be/vrtnu/a-z/trapped/1/trapped-s1a4/) in a browser (Firefox, Chromium, Midori).

dstftw commented on youtube_dl/extractor/canvas.py:
```diff
@@ -110,10 +104,153 @@ def _real_extract(self, url):
+    def _login(self):
+        username, password = self._get_login_info()
+        if username is None:
+            self.raise_login_required()
```
This is not true at the moment.

dstftw commented on youtube_dl/extractor/canvas.py:
```diff
@@ -110,10 +104,153 @@ def _real_extract(self, url):
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        display_id = mobj.group('id')
```
`self._match_id`.

mrBliss commented:
I have pushed a new commit addressing your comments except the comment(s) related to authentication, as I cannot reproduce it.

midas02 commented:
@mrBliss Could you add some code to catch links terminating in ".relevant"? The VRT uses them to redirect to the latest episode of a series. Catching them would allow one to automatically download the latest available episode.
E.g.: https://www.vrt.be/vrtnu/a-z/sportweekend.relevant/
The thing is that the current logic doesn't work for them. Their json has a keyword "url" which holds the page to redirect to. So it would need to be detected first, and redirected.
I could analyse it and propose some code if you wouldn't have time for it.

MichaelDeBoey commented:
That would indeed be a nice addition @midas02! :-)

But I think we could already merge this one and add it later?
Otherwise we keep adding extra features to the extractor without merging it to the master branch... 😞

mrBliss commented:
@midas02 Easy to add, see the last commit

mrBliss commented:
@dstftw can you look at this, if you have some time to spare?

mrBliss commented:
@dstftw do you have some time to look at this again? I would like your opinion on my last commit (f07542c)

brechtm commented:
The extractor is no longer working for me:

```
$ /Users/brechtm/Documents/Code/youtube-dl/youtube-dl --verbose --username ****** --password ****** https://www.vrt.be/vrtnu/a-z/allah-in-europa/2017/allah-in-europa-s2017a2-bosnie-hongarije/
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--verbose', u'--username', u'PRIVATE', u'--password', u'PRIVATE', u'https://www.vrt.be/vrtnu/a-z/allah-in-europa/2017/allah-in-europa-s2017a2-bosnie-hongarije/']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2017.06.12
[debug] Python version 2.7.13 - Darwin-16.7.0-x86_64-i386-64bit
[debug] exe versions: ffmpeg 3.2.4, ffprobe 3.2.4
[debug] Proxy map: {}
[VrtNU] Logging in
[VrtNU] Requesting a token
[VrtNU] allah-in-europa-s2017a2-bosnie-hongarije: Downloading webpage
[VrtNU] allah-in-europa-s2017a2-bosnie-hongarije: Downloading JSON metadata
ERROR: Unable to download JSON metadata: HTTP Error 404: Not Found (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "/Users/brechtm/Documents/Code/youtube-dl/youtube-dl/youtube_dl/extractor/common.py", line 502, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/Users/brechtm/Documents/Code/youtube-dl/youtube-dl/youtube_dl/YoutubeDL.py", line 2129, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 435, in open
    response = meth(req, response)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 548, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 473, in error
    return self._call_chain(*args)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 556, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
```

mrBliss commented:
I have updated this PR in light of 117589dfa2ac32566dc705ee7e712139105e6dd8. Because of the refactoring in the aforementioned commit, rebasing became too complex, so I squashed the commits first.

@dstftw Can you take a look at this? I know of several people (myself included) that have been using this PR for over 6 months. It would be nice if it finally got merged.

dstftw commented on youtube_dl/extractor/canvas.py:
```diff
@@ -166,3 +172,142 @@ def _real_extract(self, url):
+            # There is only one entry, but with an unknown key, so just get
+            # the first one
+            video_id = list(json.values())[0].get('videoid')
+            print(video_id)
```
All debug code should be removed.

dstftw commented on youtube_dl/extractor/canvas.py:
```diff
@@ -166,3 +172,142 @@ def _real_extract(self, url):
+        error_message = auth_info.get('errorDetails') or auth_info.get('errorMessage')
+        if error_message:
+            raise ExtractorError(
+                'Unable to login: %s' % error_message, expected=True)
```
This duplicates code from `MedialaanIE`.

mrBliss replied:
Thanks for the feedback! You're right, there is some code in common with `MedialaanIE`. How should I address this? Move the common code to `common.py`? (Let one inherit from the other is technically possible, but not my preference)

dstftw replied:
For now extract it in a base IE class.

mrBliss replied:
See 5ed058f for a first try.

dstftw commented on youtube_dl/extractor/canvas.py:
```diff
@@ -166,3 +172,142 @@ def _real_extract(self, url):
+        # We are dealing with a '../<show>.relevant' URL
+        redirect_url = json.get('url')
+        if redirect_url:
+            return self._real_extract('https:' + redirect_url)
```
Should be delegated via `url_result`.

dstftw commented on youtube_dl/extractor/canvas.py:
```diff
@@ -166,3 +172,142 @@ def _real_extract(self, url):
+
+        return {
+            '_type': 'url_transparent',
+            'url': 'https://mediazone.vrt.be/api/v1/%s/assets/%s' % ('vrtvideo', video_id),
```
`vrtvideo` should be hardcoded.

mrBliss commented:
Thanks!

hubiqs commented:
@mrBliss is there any chance I can interest you in https://github.com/rg3/youtube-dl/issues/15280 ? I would very much appreciate added support for VrtNU livestreams. Thank you for all the work you've done here already!
